Fixed `#15759` - Excluded formset fields by per-object permissions

artirix1927 opened the pull request:
ticket-15759

Fixed handling of list_editable fields in the Django admin for objects with per-object permissions. Updated changelist formset logic to remove editable fields for objects the user cannot change, ensuring that form submissions only update permitted objects and preventing permission-related crashes.

#### Checklist
- [x] This PR targets the `main` branch. <!-- Backports will be evaluated and done by mergers, when necessary. -->
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [ ] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

jacobtylerwalls reviewed (commented):
Thanks @artirix1927 👍

jacobtylerwalls reviewed (commented):
Nice work! Left a few polish comments.

artirix1927 commented on django/contrib/admin/options.py:
```diff
@@ -2108,44 +2152,20 @@ def changelist_view(self, request, extra_context=None):
-                        "name": model_ngettext(self.opts, changecount),
-                    }
-                    self.message_user(request, msg, messages.SUCCESS)
+                self._save_formset(request, cl.formset)
```
so you want me to create another pr first with just factoring it out before continuing with this one?

jacobtylerwalls replied:
Two commits on this PR is fine, but 2 PRs is fine also, whatever's easier for you.

Commit:
Refs #15759 -- Factored out _save_formset() in ModelAdmin.

Commit:
Fixed #15759 -- Excluded fields by per-object permissions for ModelAdmin.list_editable.

Instead of going over all objects in a queryset and filtering
by user permissions, added skipping while saving the formset
so there is no need to refetch objects again.

jacobtylerwalls reviewed (approved):
Thanks for the updates! 🎉

timgraham commented on tests/admin_views/admin.py:
```diff
@@ -377,6 +377,17 @@ def has_add_permission(self, request):
+    list_editable = ("gender", "alive")
+    ordering = ("id",)
+
+    def has_change_permission(self, request, obj=None):
+        if obj is None:
+            return True
+        return obj.id % 2 == 0
```
This is brittle since it relies on the database assigning IDs of a certain value. The `Person` objects for this test [aren't created with an explicit id](https://github.com/django/django/blob/526b548cfb9c8a02ea2b7ae064ef3b795305d51a/tests/admin_views/tests.py#L4633-L4635) while [the `RowLevelChangePermissionModel` objects are](https://github.com/django/django/blob/526b548cfb9c8a02ea2b7ae064ef3b795305d51a/tests/admin_views/tests.py#L2933-L2936). However, it would be better to use a different field in both cases so that the tests don't need to be edited for a database like MongoDB which doesn't use integer primary keys. Shall we create a ticket to update both?

jacobtylerwalls replied:
Great idea.

artirix1927 replied:
Is it okay though to use alive field (or any other ) for testing purposes and set one of the persons as alive=False?
If so I can fix it for both cases in a couple days.

timgraham replied:
Try it out and see if you run into any issues. Ping me on your PR and I'll be happy to test it out on a couple of third-party databases. Thanks.

artirix1927 replied:
Can you please revert this pr, I am really sorry this happened. So it does not let the user change the object he has no perms to as it should, and it also does let the user change the objects he does have permission to. But there is one ugly stage in between. I figured it out when I was working on tests refactoring.
I already fixed it, I just will have to make a new pr with a fix for the test and the feature itself or I can do just a new pr commit with the fix if you wish.

jacobtylerwalls replied:
Tell us what you know so far, and then maybe we can suggest next steps. Do you already have a test case showing something amiss?

artirix1927 replied:
Well, I guess my previous test case was a miss or I am missing a test case. So what happened is that I forgot to exclude the fields that have no permission from the validation. So whenever a user tried to save it would ask to fill all the fields. But even if you fill them it would not save the data. 


<img width="1458" height="653" alt="scr2" src="https://github.com/user-attachments/assets/14f57fff-6970-4353-ba92-504a01f90296" />
<img width="1619" height="482" alt="scr3" src="https://github.com/user-attachments/assets/052f12ee-3564-4294-9466-c55242c443f4" />

artirix1927 replied:
def _get_formset_with_permissions(self, request, queryset, data=None, files=None):
        """
        Construct a changelist formset, and remove list_editable fields
        for objects the user cannot change.
        """
        FormSet = self.get_changelist_formset(request)
        
        if data is not None:
            formset = FormSet(data, files, queryset=queryset)
        else:
            formset = FormSet(queryset=queryset)

        for form in formset.forms:
            if not self.has_change_permission(request, form.instance):
                for field_name in self.list_editable:
                    form.fields.pop(field_name, None)

So I fixed it by adjusting this function and taking the formset from this function for POST formset as well.
For the brittle test with usage of `obj.id % 2` I changed it to return `obj.alive` and everything works fine. As well as I added alive fields in my POST data for the test because it was missing.

artirix1927 replied:
So as I understand for now my initial test was not perfectly simulating the real case because in the POST data I put data for all three of the persons, which in real case would be only the ones that we have permission to. Then in the post data there would be missing data for 1/3 person (since we have no permissions to change it) and it would not validate. But because in the test we had data for all 3 it, the full data would get passed to the form (post handler) and it would be saved and validated successfully. But because for the post handler I didn't use special formset (that excludes fields we have no permission to), in the real world with missing data it would throw an error. So by using special formset with permissions for post handler as well I fix that bug.

timgraham replied:
Follow up PR: https://github.com/django/django/pull/21179
